core/src/main/java/hudson/model/TopLevelItemDescriptor.java

```diff
@@ -114,6 +131,85 @@ public String getDisplayName() {
     }
 
     /**
+     * A description of this kind of item type. This description can contain HTML code but it is recommend to use text plain
+     * in order to avoid how it should be represented.
+     *
+     * This method should be called in a thread where Stapler is associated, but it will return an empty string.
+     *
+     * @return A string, by default the value from newInstanceDetail view is taken.
+     *
+     * @since TODO
+     */
+    @Nonnull
+    public String getDescription() {
+        Stapler stapler = Stapler.getCurrent();
+        if (stapler != null) {
+            try {
+                WebApp webapp = WebApp.getCurrent();
+                MetaClass meta = webapp.getMetaClass(this);
+                Script s = meta.loadTearOff(JellyClassTearOff.class).findScript("newInstanceDetail");
+                if (s == null) {
+                    return "";
+                }
+                DefaultScriptInvoker dsi = new DefaultScriptInvoker();
+                StringWriter sw = new StringWriter();
+                XMLOutput xml = dsi.createXMLOutput(sw, true);
+                dsi.invokeScript(Stapler.getCurrentRequest(), Stapler.getCurrentResponse(), s, this, xml);
+                return sw.toString();
+            } catch (Exception e) {
+                LOGGER.warning(e.getMessage());
+                return "";
+            }
+        } else {
+            return "";
+        }
+    }
+
+    /**
+     * Used to categorize this kind of item type. @see {@link ItemCategory}
+     *
+     * @return A string with the category identifier, {@link ItemCategory.UncategorizedCategory#getId()} by default.
+     *
+     * @since TODO
+     */
+    @Nonnull
+    public String getCategoryId() {
+        return ItemCategory.UncategorizedCategory.ID;
+    }
+
+    /**
+     * Represents a file path pattern to get the Item icon in different sizes.
+     *
+     * For example: plugin/plugin-shortname/images/:size/item.png, where {@code :size} represents the different
+     * icon sizes used commonly in Jenkins project: 16x16, 24x24, 32x32 or 48x48
+     *
+     * @return A string or null if it is not defined.
+     *
+     * @since TODO
+     */
+    @CheckForNull
+    public String getIconFilePathPattern() {
+        return null;
+    }
+
+    /**
+     * An icon file path associated to a specific size.
+     *
+     * @param size A string with values that represent the common sizes: 16x16, 24x24, 32x32 or 48x48
+     *
+     * @return A string or null if it is not defined.
+     *
+     * @since TODO
+     */
+    @CheckForNull
+    public String getIconFilePath(String size) {
```

Comment:
@jglick 

> (Still not sure why this method exists, since it seems unused.)

This method makes sense because exists this one `getIconFilePathPattern()`. You can see an use case in this PR.

> should probably be `final` to emphasize that overriding this method will not accomplish anything on its own.

This current implementation is flexible. You can override `getIconFilePath()` and using the default implementation of `getIconFilePathPattern()`. For example:

```
@CheckForNull
public String getIconFilePath(String size) {
    return "plugin/maven-plugin/images/my-icon.png";
}
```